pkg/ruler/ruler.go

```diff
@@ -171,36 +172,40 @@ func NewRuler(cfg Config, engine *promql.Engine, queryable storage.Queryable, d
 		workers[i] = newWorker(ruler)
 	}
 
-	ruler.workers = workers
-	go ruler.run()
-
-	return ruler, nil
-}
-
-// Run the ruler.
-func (r *Ruler) run() {
-	go r.scheduler.Run()
-	for _, w := range r.workers {
-		go w.Run()
+	go ruler.scheduler.Run()
+	for i := range workers {
+		// initialize each worker in a function that signals when
+		// the worker has completed
+		go func(w worker) {
+			ruler.workers.Add(1)
```

Comment:
`Add` should always be called outside of the goroutine being created. Otherwise, you can end up with some weird race conditions where workers could be created after `Wait()` returns.

Reply:
The race condition I am describing is probably also why the tests are failing.

Reply:
Good catch, it should be good now